WIP: Development Environment: Docker and docker-compose setup

aantonop opened the pull request:
Initial attempt at Dockerfile for standardized dev environment

See #158

Installs dev tools, bitcoind, c-lightning, lnd

0xmountaintop committed:
add "Ligtning Messages" section

0xmountaintop committed:
fix typos

0xmountaintop committed:
format table

0xmountaintop committed:
add "it's OK to be odd" rule

0xmountaintop committed:
fix typos

0xmountaintop committed:
add "ignore additional data" rule

0xmountaintop committed:
lower dowm header levels

aantonop committed:
Development Environment: Dockerfile draft

Installs dev tools, bitcoind, c-lightning, lnd

aantonop committed:
Development Environment: Dockerfile draft

Installs dev tools, bitcoind, c-lightning, lnd

aantonop committed:
Added eclair 0.3.3

aantonop committed:
Added eclair 0.3.3

renepickhardt commented:
what about Rust Lightning (becoming LDK by square crypto), Electrum and https://github.com/nayutaco/ptarmigan ?

aantonop committed:
Separate dockerfiles and docker-compose

aantonop committed:
Separate dockerfiles and docker-compose

bolatovumar commented:
@renepickhardt isn't Rust Lightning still not really a full implementation of the LN protocol and way behind LND/c-lightning/Eclair? It probably will be more feature-complete by the time this book is published though.

fguisso commented on code/docker/bitcoind-base/Dockerfile:
```diff
@@ -0,0 +1,36 @@
+FROM ubuntu:18.04
```
Golang have some good official images than ubuntu https://hub.docker.com/_/golang
With Debian, windowsservercore or alpine, you don't need to install the binaries and configs

0xmountaintop replied:
agree. and this can reduce the docker image size.

aantonop committed:
Use golang docker image for lnd_base

aantonop committed:
Use golang docker image for lnd_base

aantonop committed:
Use openjdk:11 docker image for eclair

aantonop committed:
Use openjdk:11 docker image for eclair

aantonop committed:
bitcoind regtest docker image

aantonop committed:
bitcoind regtest docker image

St333p commented:
What about [lighter](https://gitlab.com/inbitcoin/lighter)?

It would allow users to have a uniform interface across implementation (c-lightning, eclair, lnd supported; electrum coming soon and ptarmigan in roadmap) at least for the most common operations. Of course one still needs to directly talk to c-lightning to use plugins, but that is gonna be 5% of readers. 

I might also find some time to write about it and to fix the _compose_ we are using for internal tests for publication if there is interest in this regard.

korynewton committed:
fix glossary typos (#167)

Commit:
minor improvements and clarifications (#170)

- minor clarifications
- incorrect use of plural
- repetitive use of "this way" --> this fashion
- clarifying "each time" --> each payment
- added detail that LN txs are not stored permanently
- repetitive use of "explanation" --> clarification
- seeing --> revisiting
- etc.

Commit:
clarifications and additions (#173)

* clarifiactions and additions

some modifications to improve, clarify

some additions to add more details where appropriate

... lightweight Bitcoin client (commonly referred to as Simplified Payment Verification (SPV)) to partially validate the correctness of their blockchain. --> partially? Probably you mean "indirectly" as you delegate to a third party. Also SPVs clients don't have a blockchain, so why "their blockchain"? --> to reduce this confusion --> lightweight Bitcoin client (commonly referred to as Simplified Payment Verification (SPV)) to interact with the blockchain

... lightning wallets ... --> ... Lightning wallets ...

centralized and added more details to definition of non-custodial wallet

* Update 02_getting_started.asciidoc

added the sentince back in

Co-authored-by: Rene Pickhardt <rene@rene-pickhardt.de>

Commit:
improved definition of LN (#177)

- removed incorrect comma in front of "as well as"
- added "both" to make it stronger that LN has 2 main functionalities
- "... on the network." leaves it undefined. The reader might ask to which netowrk it refers to? To the before mentioned netowrk of payment-channels? Added "Bitcoin" to clarify.

Commit:
added a few words to improve definition paragraph (#178)

- consistent use of "partner"
- added "Bitcoin" to "Bitcoin smart contract" because there are many people out there that have heard the term "smart contract" exclusively in the context of Ethereum. It might be worth while to underline that Bitcoin also has its form of smart contracts. This should eliminate any possible doubts.
- imply that earlier states are no longer valid, added "invalidated" 
- channel member is singular, so "if they try" is wrong. better "if it tries"

Commit:
minor improvements in Note (#179)

- 2-out-of-2 is more consistent (see usage above) and carries more information than 2-2
- comma after however needed

Commit:
minor improvements, corrections (#180)

- comma must follow "additionally" and "however" as they are introductory words or phrases
- same for "last but not least", comma missing
- "nothing else than", not perfect English; better "nothing but" 
- "get most use", not perfect English; better "get more benefits"

Commit:
minor corrections in Payment channels paragraphs (#181)

- "the both of you": incorrect English, better "both of you"
- "cheat on you": spoken English, better "cheat you"
- "had to be", that sounds like you are guessing, better "was"
- if you give full names of some, give full names of all 
- added year to make it more precise
- etc.

Commit:
improvements to Multisig addresses paragraphs (#182)

- as before "2-out-of-2" is more appropriate than "2-2" 
- "thus" is not what you wanted to express, "however" is what you wanted to say
- added missing comma
- etc

Commit:
minor improvements to Funding Transaction paragraphs (#183)

- consistently sticking to "2-out-of-2"
- rephrasing some sentences to make them more palatable. "obviously a lie" is a bit aggressive. 
- avoid using exaggerating terms like "very expensive", just state the facts without too many emotions

Commit:
minor improvements in "Example of a poor channel opening procedure" (#184)

- replace "committing attack" with "performing attack" to avoid that the reader incorrectly concludes that the term "commitment tx" has anything to do with "committing an attack"
- signed by, not signed from

entrey committed:
fix typos in ch1 and ch2 (#168)

* table indentation

* 'a' before consonant

* less commas

* wording

* unquote commas

* typo

* on chain -> on-chain

* need comma (IMHO)

* dot

* caps acronym

* wording

Co-authored-by: Rene Pickhardt <rene@rene-pickhardt.de>

korynewton committed:
fix typos in chapter 3 (#166)

* fix typos Chapter 3

* additional ch3 typo fixes

Co-authored-by: Rene Pickhardt <rene@rene-pickhardt.de>

0xmountaintop committed:
add Transaction Malleability (#150)

* add Transaction Malleability

* fix https://github.com/lnbook/lnbook/pull/150#discussion_r379919498

* explain why non-malleability is needed

* fix typos

renepickhardt committed:
added some calaculations about onchain scaling

renepickhardt committed:
added the fist version of the routing chapter. did no have time to do proof reading or editorial yet

renepickhardt committed:
added explainations how htlcs are being set up in payment channels to prepare routing

aantonop committed:
Merge pull request #154 from HAOYUatHZ/ln_msg

add "Ligtning Messages" section

aantonop committed:
updates

aantonop committed:
updates

aantonop committed:
Relaxed Travis to avoid breaking builds while things are in flux

The travis linter is too strict and we keep failing builds. This is a bit like taking the batteries out of the smoke alarm, or using aluminum foil to replace a fuse, or both. Everything is fine!

renepickhardt committed:
started to write about setteling htlcs

Commit:
minor improvements to chapter "Missing bits" (#202)

- repeated use of "following"
- peer-to-peer instead "peer to peer"
- minor clarifications
- etc

Commit:
minor improvements to chapter "Finding a path" (#200)

- added commas
- some sentences presented the ideas in the wrong order and the sentences were too long --> restructured sentences
- etc

Commit:
minor improvements to chapter "Delivering the payment" (#199)

- peer-to-peer better than "peer 2 peer" , in acronyms p2p is ok
- English language guidelines recommend numbers up to 12 to be spelled out. 4 --> four
- minor additions/clarifications

Commit:
minor improvements to chapter "Payment Hash" and "Additional Meta Data" (#198)

- repeated use of "also", --> furthermore, ...
- minor clarifications
- etc

Commit:
added contributor name (#195)

+1 line  :)

Commit:
Improvements to chapter "Commitment Transaction" (#188)

- replace "capacity" with "funds to reduce tech jargon
- replaced 2-2 with 2-out-of-2 for consistency and to clarify
- everyone --> singular  .. its funds .. not their funds
- "wallet" wrong, "address" better
- avoid "flaw", it might get misinterpreted, rephrased to clearly state that we present the design step-by-step
- "you probably have realized" and "hope you recognize that" puts the reader in a bind. If the reader does not see the problem we will feel "dumb" because the text implies "probably you realized". I suggest to rewrite it so, that the reader does not feel "dumb" just because he does not see the shortcut. --> see changes
- however requires comma
- avoid extremes like "rather high", use facts and measurable units, here simply "high"
- "to lose funds" incorrect, "loose" is correct
- took one long sentence and split it up into several short one making understanding a lot simpler
- etc

Commit:
improvements to chapter "Commitment Transaction", part 2 (#189)

- comma after however
- when referring to time "immediately" is better and more meaningful than "directly"
- when referring to time duration "lasts" is better than "is"
- consistency: "channel partner" better than "party"
- etc

Commit:
minor improvements to "Announcing the channel" (#190)

- added distinction between "public" and "private" channel, added the 2 terms
- rename 2-2 to 2-out-of-2 for consistency and to make it clearer
- simplified a few sentences
- added a few additional clarifying words
- etc

Commit:
minor improvements to chapter "Closing the channel" (#191)

- Bitcoin: uppercase as it refers to fees on the Bitcoin protocol
- added more detail on importance of longevity of channel
- the last sentence was way too long and too complicated. restructured it.
- etc

Commit:
minor improvements to chapter "Examining the good way - mutual close" (#192)

- closing tx, shutdown tx, etc:  to avoid any confusion I took the "shutdown tx" out and simply named it the "on-chain tx"
- replaced "party" with "channel partner" for consistency
- removed "as many people think". If the book does a good job, not many people will think that ! In any case, this phrase does not help. Better to just state the facts. 
- etc

Commit:
minor improvements to chapter "Examining the bad way" (#193)

- simplified
- reduced sentence length
- comma after thus
- with respect to time "immediately" is better than "directly"
- avoid extremes, replaced "much higher" with "higher", later we specify the figure 5x anyway
- "additional" better than "more" 
- misc clarifications
- some additions
- etc

Commit:
minor improvements to chapter "Examining the ugly way" (#194)

- again, avoid extreme terms like "really high" that cannot be quantified.
- you do not gain the entire channel capacity, gain means "benefit", if 9 mBTC out of the 10 mBTC channel capacity were yours anyway, you would GET 10 mBTC but only GAIN 1 mBTC.
- additions
- mention of game theory
- etc

Commit:
minor improvements to chapter "Comparison with Bitcoin" part 1

- no commas when subject is the same on both parts of sentence
- nothing seems to be gained by introducing concept of "traffic" --> removed it
- instead explained what the scarce resources are
- more explanations
- etc.

renepickhardt committed:
fix htlc typo that came in through find and replace

Commit:
minor improvements to chapter "Comparison with Bitcoin" part 2

- more details
- etc

Commit:
addition to chapter "Comparison with Bitcoin" part 6

- added new chapter

HugoDoyon committed:
Terrabyte sould be Terabyte (#211)

Typo

HugoDoyon committed:
typo (#212)

inormation should be in"f"ormation

HugoDoyon committed:
Learn and comma

Laern instead of Lear

HugoDoyon committed:
Typos

challanges & whith should be written: challenges and with

HugoDoyon committed:
feasable?

feasible (typo)

Roasbeef committed:
Merge pull request #204 from 8go/patch-29

minor improvements to chapter "Comparison with Bitcoin" part 1

Roasbeef committed:
Merge pull request #205 from 8go/patch-30

minor improvements to chapter "Comparison with Bitcoin" part 2

Roasbeef committed:
Merge pull request #209 from 8go/patch-35

addition to chapter "Comparison with Bitcoin" part 6

Roasbeef committed:
Merge pull request #214 from HugoDoyon/patch-3

Learn and comma

Roasbeef committed:
Merge pull request #215 from HugoDoyon/patch-4

Typos

Roasbeef committed:
Merge pull request #216 from HugoDoyon/patch-5

feasable?

Commit:
ch1+ch3: fix small typos

Commit:
ch3: minor improvements to chapter "Thoughts about Trust" 


- principal --> Major
- I would not call cost a risk. The opening fee IMHO is not a risk, it is simply a cost.
- commas
- in reference to time "immediately" better than "directly"
- added another example, to explain when not to create a channel

Roasbeef committed:
chp2p: add chapter outline for p2p chapter

Roasbeef committed:
ch-chan-link: add chapter overview

Roasbeef committed:
ch-htlc-forwarding: add chapter overview

Roasbeef committed:
ch-channel-graph: add chapter overview

Roasbeef committed:
ch-path-finding: add chapter overview

Roasbeef committed:
ch-e2e: add chapter overview

Roasbeef committed:
fixup! chp2p: add chapter outline for p2p chapter

HugoDoyon committed:
then

Instead of "theh", should be writtent "then"

HugoDoyon committed:
Implicitly (Minor correction)

"implicitely" should be "implicitly"

HugoDoyon committed:
Minor correction when"S"

When does not need "s"

HugoDoyon committed:
Minor correction

"anormality" should be written "abnormality"

aantonop committed:
Merge pull request #219 from Roasbeef/chapter-outlines

multi: add skeleton chapter outlines

aantonop committed:
Merge pull request #220 from HugoDoyon/patch-6

then

aantonop committed:
Merge pull request #221 from HugoDoyon/patch-7

Implicitly (Minor correction)

aantonop committed:
Merge pull request #222 from HugoDoyon/patch-9

Minor correction when"S"

aantonop committed:
Merge branch 'develop' into patch-10

aantonop committed:
Merge pull request #223 from HugoDoyon/patch-10

Minor correction

aantonop committed:
New README with status

aantonop committed:
README - Status by section and wordcount

oneeman committed:
Fix 'loose' vs 'lose' mistakes

renepickhardt committed:
saving stub for onion routing. being stuck at deciding which detail level I should aim. It is tempting to include shared secrete generation for the paths including computation of blinding factors of ephemeral keys. While I believe it would be highly instructional and edcuational it is kind of a random deep dive into standard cryptographic methods

HugoDoyon committed:
publicly insteat of pubicly  

publicly insteat of pubicly

HugoDoyon committed:
lose instead of loose

lose instead of loose

HugoDoyon committed:
payment instead of payement

payment instead of payement

HugoDoyon committed:
crucially instead of crucually 

crucually should be written crucially instead

HugoDoyon committed:
channel typo: chennel

channel instead of chennel

aantonop committed:
Merge pull request #227 from HugoDoyon/patch-13

payment instead of payement

aantonop committed:
Merge pull request #229 from HugoDoyon/patch-15

channel typo: chennel

aantonop committed:
Merge pull request #228 from HugoDoyon/patch-14

crucially instead of crucually

HugoDoyon committed:
netowrk should be network

netowrk should be network

aantonop committed:
Merge pull request #224 from oneeman/loose-lose-fix

Fix 'loose' vs 'lose' mistakes

aantonop committed:
Merge pull request #225 from HugoDoyon/patch-11

publicly insteat of pubicly

aantonop committed:
Merge pull request #226 from HugoDoyon/patch-12

lose instead of loose

aantonop committed:
Merge pull request #230 from HugoDoyon/patch-16

netowrk should be network

gin committed:
Clarification, and define term to be consistent

Definition for "c-lightning" to be consistent with "Eclair".
Correct typo.

gin committed:
Fill "encoding" definition

0xmountaintop commented on code/docker/c-lightning-base/Dockerfile:
```diff
@@ -0,0 +1,25 @@
+FROM ubuntu:18.04
```
we use bitcoind's prebuilt docker image but we build lightningd image by ourselves. will it be a good idea to also use   c-lightning/lnd's docker image?

lnd's: https://github.com/lightningnetwork/lnd/blob/master/docker/README.md

it also provide a good tutorial on how nodes interact with each other, and [how to run the image](https://github.com/lightningnetwork/lnd/blob/master/docs/DOCKER.md)

0xmountaintop replied:
I just search.   seems  there's no official  docker images of   c-lightning /  lnd    published on docker hub.  I believe that's the reason @aantonop choose to build it by ourselves.

alternatives approaches can be:

+ use https://hub.docker.com/r/cdecker/lightning-ci/dockerfile
+ or maybe @Roasbeef  can publish a docker image onto docker hub on behalf of lightninglabs?

0xmountaintop commented on code/docker/devenv/Dockerfile:
```diff
@@ -0,0 +1,65 @@
+ENV PATH $PATH:$GOPATH/bin
+
+# LND
+RUN go get -d github.com/lightningnetwork/lnd
```
will it be good if we specify a version (using commit hash or ,,,).   

My concern is that,    there may be some breaking changes, and the commands may become incompatible. for example, https://en.bitcoin.it/wiki/Original_Bitcoin_client/API_calls_list these are deprecated.

gin committed:
Fill "Transport Layer" definition, add punctuations, correct typos

gin committed:
Add name to preface

gin committed:
Update . to , for consistency

aantonop committed:
Merge pull request #233 from gin/glossary

Edits to glossary

aantonop committed:
Merge pull request #234 from gin/ch01

Update . to , for consistency

aantonop committed:
Merge remote-tracking branch 'aantonop_fork/devenv' into devenv
